### PR TITLE
Add Hashing support to PublicKey and Transaction

### PR DESCRIPTION
**Pull request by AndrejMitrovic**

### CODE REVIEW

File: source/agora/common/Transaction.d [hunk 1]
```diff
@@ -54,6 +55,32 @@ public struct Transaction
...
+        foreach (output; this.outputs)
+            hashPart(output, dg);
+    }
+}
+
+///
+unittest
+{
+    Transaction tx;
+    auto hash = hashFull(tx);
```
Comment (AndrejMitrovic) [inline source/agora/common/Transaction.d:83]: hello missing test

**Comment by AndrejMitrovic** [comment]
I think I am missing some tests. I will look into it more on Monday.

File: source/agora/common/crypto/Key.d [hunk 2]
```diff
@@ -160,6 +161,31 @@ public struct PublicKey
...
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        dg(this.data[]);
+    }
+
+    ///
+    unittest
```
Comment (Geod24) [inline source/agora/common/crypto/Key.d:179]: Place attributes on the unittest

File: source/agora/common/crypto/Key.d [hunk 3]
```diff
@@ -160,6 +161,31 @@ public struct PublicKey
...
+    {
+        immutable address = `GDD5RFGBIUAFCOXQA246BOUPHCK7ZL2NSHDU7DVAPNPTJJKVPJMNLQFW`;
+        PublicKey pubkey = PublicKey.fromString(address);
+
+        auto hash = hashFull(pubkey);
+        auto exp_hash = Hash("0xdb3b785e31c05c9383f498aa12a5c054fcb6f99b20bf7ff25fdb24d5fe6d1bb5768f1821de9d5d31faa7ebafc79837ba32dc4774bffea9918411d0c4c8ac403c");
```
Comment (Geod24) [inline source/agora/common/crypto/Key.d]: Break it down at 80 columns ?

**Comment by AndrejMitrovic** [comment]
Updated to include attributes (where possible), and broke apart strings at the 80-column limit.

File: source/agora/common/crypto/Key.d [hunk 4]
```diff
@@ -15,6 +15,7 @@ module agora.common.crypto.Key;
 
 import agora.common.Data;
 import agora.common.crypto.Crc16;
+import agora.common.Hash;
```
Comment (Geod24) [inline source/agora/common/crypto/Key.d:18]: Do you still need this commit ?

Comment (AndrejMitrovic) [reply source/agora/common/crypto/Key.d:18]: I am trying to think if I don't.

Comment (Geod24) [reply source/agora/common/crypto/Key.d:18]: Yes you do 🤦‍♂

**Review by Geod24 — APPROVED** [review]

File: source/agora/common/crypto/Key.d [hunk 5]
```diff
@@ -160,6 +161,36 @@ public struct PublicKey
...
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        dg(this.data[]);
+    }
+
+///
+unittest
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: aaaaaaaaah alignment!!!

**Comment by AndrejMitrovic** [comment]
Pushed fix for code alignment.

**Comment by Geod24** [comment]
> Pushed fix for code alignment.

In the wrong commit tho

**Comment by AndrejMitrovic** [comment]
My mistake, it's because of that extra commit I included. It should be fixed now.

**Pull request merged**